server/src/main/java/io/crate/execution/engine/collect/files/LineProcessor.java

```diff
@@ -21,60 +21,124 @@
+        source.moveToStart();
+        firstLine = true;
+    }
+
+    private boolean readFirstLine(URI currentUri, String line) throws IOException {
+        if (isCSV(inputFormat, currentUri)) {
+            csvLineParser = new CSVLineParser(parserProperties, targetColumns);
+            inputFormat = InputFormat.CSV;
+            if (parserProperties.fileHeader()) {
+                csvLineParser.parseHeader(line);
+                return true;
+            }
+        } else {
+            inputFormat = InputFormat.JSON;
         }
+        return false;
     }
 
-    void startWithUri(URI currentUri) {
-        lineContext.resetCurrentLineNumber();
-        lineContext.currentUri(currentUri);
+    private byte[] getByteArray(String line, long rowNumber) throws IOException {
+        if (inputFormat == InputFormat.CSV) {
+            return parserProperties.fileHeader() ?
+                csvLineParser.parse(line, rowNumber) : csvLineParser.parseWithoutHeader(line, rowNumber);
+        } else {
+            return line.getBytes(StandardCharsets.UTF_8);
+        }
     }
 
-    void readFirstLine(URI currentUri, InputFormat inputFormat, BufferedReader currentReader) throws IOException {
-        lineParser.readFirstLine(currentUri, inputFormat, currentReader);
+    private static boolean isCSV(FileUriCollectPhase.InputFormat inputFormat, URI currentUri) {
+        return (inputFormat == FileUriCollectPhase.InputFormat.CSV) || currentUri.toString().endsWith(".csv");
     }
 
-    public void process(String line) throws IOException {
-        lineContext.incrementCurrentLineNumber();
-        lineContext.resetCurrentParsingFailure(); // Reset prev failure if there is any.
-        byte[] jsonByteArray = lineParser.getByteArray(line, lineContext.getCurrentLineNumber());
-        lineContext.rawSource(jsonByteArray);
+    @Override
+    public boolean moveNext() {
+        try {
+            while (source.moveNext()) {
+                LineCursor cursor = source.currentElement();
+                String line = cursor.line();
+                if (line == null) {
+                    assert cursor.failure() != null : "If the line is null, there must be a failure";
+                    return true;
+                }
+                if (firstLine) {
+                    firstLine = false;
+                    if (readFirstLine(cursor.uri(), line)) {
+                        continue;
+                    }
+                }
+                try {
+                    byte[] json = getByteArray(line, cursor.lineNumber());
+                    lineContext.resetCurrentParsingFailure();
+                    lineContext.rawSource(json);
+                } catch (Throwable parseError) {
+                    lineContext.setCurrentParsingFailure(parseError.getMessage());
+                }
+                return true;
+            }
+            return false;
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
```

Comment:
afaik, this BatchIterator comes after `FileReadingIterator` ie this won't go into `lineContext.LineCursor.failure` (and it shouldn't)

I remember CSV parsing library throws IOExceptions on parsing (I think it's csvReader.readValues doing that) - ie we probably should qualify this as a parsing failure and call `lineContext.setCurrentParsingFailure`?